pallets/restricted-xtokens/src/lib.rs

```diff
@@ -0,0 +1,428 @@
+		/// `MultiLocation` id.
+		///
+		/// If `fee` is not high enough to cover for the execution costs in the
+		/// destination chain, then the assets will be trapped in the
+		/// destination chain
+		///
+		/// It's a no-op if any error on local XCM execution or message sending.
+		/// Note sending assets out per se doesn't guarantee they would be
+		/// received. Receiving depends on if the XCM message could be delivered
+		/// by the network, and if the receiving chain would handle
+		/// messages correctly.
+		#[pallet::call_index(3)]
+		#[pallet::weight(orml_xtokens::XtokensWeight::< T >::weight_of_transfer_multiasset(asset, dest))]
+		pub fn transfer_multiasset_with_fee(
+			origin: OriginFor<T>,
+			asset: Box<VersionedMultiAsset>,
+			fee: Box<VersionedMultiAsset>,
+			dest: Box<VersionedMultiLocation>,
+			dest_weight_limit: WeightLimit,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin.clone())?;
+			let multi_asset: MultiAsset = (*asset.clone())
+				.try_into()
+				.map_err(|()| orml_xtokens::Error::<T>::BadVersion)?;
+			let fee_asset: MultiAsset = (*fee.clone())
+				.try_into()
+				.map_err(|()| orml_xtokens::Error::<T>::BadVersion)?;
+			let destination: MultiLocation = (*dest.clone())
+				.try_into()
+				.map_err(|()| orml_xtokens::Error::<T>::BadVersion)?;
+
+			ensure!(
+				T::PreTransfer::check(TransferEffects::TransferMultiAssetWithFee {
+					sender,
+					destination,
+					asset: multi_asset,
+					fee_asset
+				}),
+				Error::<T>::RestrictionTriggered
+			);
+
+			orml_xtokens::Pallet::<T>::transfer_multiasset_with_fee(
+				origin,
+				asset,
+				fee,
+				dest,
+				dest_weight_limit,
+			)
+		}
+
+		/// Transfer several currencies specifying the item to be used as fee
+		///
+		/// `dest_weight_limit` is the weight for XCM execution on the dest
+		/// chain, and it would be charged from the transferred assets. If set
+		/// below requirements, the execution may fail and assets wouldn't be
+		/// received.
+		///
+		/// `fee_item` is index of the currencies tuple that we want to use for
+		/// payment
+		///
+		/// It's a no-op if any error on local XCM execution or message sending.
+		/// Note sending assets out per se doesn't guarantee they would be
+		/// received. Receiving depends on if the XCM message could be delivered
+		/// by the network, and if the receiving chain would handle
+		/// messages correctly.
+		#[pallet::call_index(4)]
+		#[pallet::weight(orml_xtokens::XtokensWeight::< T >::weight_of_transfer_multicurrencies(currencies, fee_item, dest))]
+		pub fn transfer_multicurrencies(
+			origin: OriginFor<T>,
+			currencies: Vec<(T::CurrencyId, T::Balance)>,
+			fee_item: u32,
+			dest: Box<VersionedMultiLocation>,
+			dest_weight_limit: WeightLimit,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin.clone())?;
+			let destination: MultiLocation = (*dest.clone())
+				.try_into()
+				.map_err(|()| orml_xtokens::Error::<T>::BadVersion)?;
```

Comment:
Wondering if using errors from `orml_xtokens` can confuse a future us trying to find an error thinking the error comes from `orml_xtokens` pallet

Reply:
Maybe it can be nested under `restricted_xtokens` error?

Reply:
The palllet itself is hard-coupled to `orml-xtokens` and the error will also happen in `orml_xtokens`. But I can of course make a dedicated `BadVersion` for restricted-xtokens.

Reply:
Maybe something like: `restricted_xtokens::Error::XTokens(orml_xtokens::Error)` could work?

If not, I would say to make a dedicated version

Reply:
Of course, but that is not the error. It would be a dedicated copy error of the one in xtokens

Reply:
Could we do the same for `AssetIndexNonExistent`? To align it